diff -Naur aide-0.16.orig/src/db_disk.c aide-0.16/src/db_disk.c
--- aide-0.16.orig/src/db_disk.c	2024-03-11 16:45:06.594013966 -0400
+++ aide-0.16/src/db_disk.c	2024-03-11 16:45:06.584013966 -0400
@@ -209,7 +209,7 @@
 		fullname=malloc((conf->root_prefix_length+2)*sizeof(char));
 		strcpy(fullname, conf->root_prefix);
 		strcat (fullname, "/");
-		if (!get_file_status(&fullname[conf->root_prefix_length], &fs)) {
+		if (!get_file_status(fullname, &fs)) {
 		add = check_rxtree (&fullname[conf->root_prefix_length], conf->tree, &attr, fs.st_mode);
 		error (240, "%s match=%d, tree=%p, attr=%llu\n", &fullname[conf->root_prefix_length], add,
 					 conf->tree, attr);
@@ -255,7 +255,7 @@
 		   If not call, db_readline_disk again...
 		 */
 
-		if (get_file_status(&fullname[conf->root_prefix_length], &fs)) {
+		if (get_file_status(fullname, &fs)) {
 		    free (fullname);
 		    goto recursion;
 		}
